sdks/python/apache_beam/io/gcp/gcsio.py

```diff
@@ -262,33 +269,18 @@ def delete_batch(self, paths):
              succeeded or the relevant exception if the operation failed.
     """
     final_results = []
-    s = 0
-    if not isinstance(paths, list): paths = list(iter(paths))
-    while s < len(paths):
-      if (s + MAX_BATCH_OPERATION_SIZE) < len(paths):
-        current_paths = paths[s:s + MAX_BATCH_OPERATION_SIZE]
-      else:
-        current_paths = paths[s:]
-      current_batch = self.client.batch(raise_exception=False)
-      with current_batch:
-        for path in current_paths:
-          bucket_name, blob_name = parse_gcs_path(path)
-          bucket = self.client.bucket(bucket_name)
-          bucket.delete_blob(blob_name)
-
-      for i, path in enumerate(current_paths):
-        error_code = None
-        resp = current_batch._responses[i]
-        if resp.status_code >= 400 and resp.status_code != 404:
-          error_code = resp.status_code
-        final_results.append((path, error_code))
-
-      s += MAX_BATCH_OPERATION_SIZE
-
+    for path in paths:
+      error_code = None
+      try:
+        self.delete(path)
```

Comment:
I am hesitating whether if worth to undo the batch operation, which may have efficiency regression. In a FileIO write, batch copy and delete happens at the end of a pipeline (not the most resource restraing part). I suggest we leave batch copy and delete as is.

Reply:
One problem of using the batch operation of gcs client library is that it does not support retries. In terms of performance, maybe @andrewsg can shed some lights here?

Reply:
If so, I think we do not need to add throttling metrics counter to batch copy and delete. By construct if they do not retry then there is no wait time due to throttling.

Alternatively we can check the return code of the batch process, if a resource exahusted request indeed returns some http error, we can treat the whole blocking operation time as a throttled time and report it with the throttling counter

Reply:
I think retrying on batch operation is useful too, but I agree with your concern of performance regression. I think an alternative way is to keep both the old and new implementation. We can call the old way (fast code path) first, if there is anything wrong, we then call the new way (slow code path) which increments the throttling counter.

Reply:
Ok. Since we have some concern on the performance on new implementation of copy_batch and delete_batch, I revert the changes there but also add a notes in the docstring to make it clear that these two operations will not be retried if any exception occurs.